### PR TITLE
fix(GCS+gRPC): correctly resume ranged reads

### LINKED ISSUE #14494: GCS+gRPC incorrectly updates the limit for resumed ranged reads
Basically this code is wrong:

https://github.com/googleapis/google-cloud-cpp/blob/eb2caccb6ebfb14f27cea6ee1f94680966fcdf67/google/cloud/storage/internal/grpc/object_request_parser.cc#L392-L394

The fix is to reduce the limit by the number of bytes received.  I will send a PR shortly, I just want a bug report in case somebody else runs into this.

### PR DESCRIPTION
**Pull request by coryan**
Fixes #14494

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/14495)

### CODE REVIEW

File: google/cloud/storage/internal/grpc/object_request_parser.cc [hunk 1]
```diff
@@ -388,9 +388,15 @@ StatusOr<google::storage::v2::ReadObjectRequest> ToProto(
   }
   if (request.HasOption<storage::ReadFromOffset>()) {
```
Comment (dbolduc) [inline google/cloud/storage/internal/grpc/object_request_parser.cc:389]: I get the math, but it's confusing to read. Consider only setting `read_offset`, `read_limit` once for any configuration of options. e.g.:

```cc
  auto const offset = request.HasOption<storage::ReadFromOffset>()
                          ? request.GetOption<storage::ReadFromOffset>().value()
                          : 0;
  if (request.HasOption<storage::ReadRange>()) {
    auto const range = request.GetOption<storage::ReadRange>().value();
    auto const start = (std::max)(range.begin, offset);
    auto const limit = range.end - start;
    if (limit <= 0) return internal::InvalidArgumentError("blah blah");

    r.set_read_offset(start);
    r.set_read_limit(limit);
  } else if (offset != 0) {
    r.set_read_offset(offset);
  }
```

**Review by dbolduc — COMMENTED** [review]
Q: why should we even allow `ReadFromOffset` and `ReadRange` together? It is only well defined if `offset == range.begin`, no?

**Comment by coryan** [comment]
> Q: why should we even allow `ReadFromOffset` and `ReadRange` together? It is only well defined if `offset == range.begin`, no?

You could argue that the intersection of `ReadFromOffset` and `ReadRange` is reasonably well defined if `offset < range.end`.

Cleverly (see the air quotes?) the resume loop only sets `ReadFromOffset`, when/if the application also sets `ReadRange` we merge both.  Not the cleanest design, but a bit tricky to (a) prohibit, given the APIs we do expose, or (b) fix without risking a break  for JSON.

Let me consider your proposed fix separately.

File: google/cloud/storage/internal/grpc/object_request_parser.cc (hunk 1 above)
Comment (dbolduc) [reply google/cloud/storage/internal/grpc/object_request_parser.cc:389]: It would not work if negative ranges are a thing. I assume `range.begin >= 0`.

Comment (coryan) [reply google/cloud/storage/internal/grpc/object_request_parser.cc]: Negative ranges are not a thing, that is not something you can send to the service.  I am not overly concerned about detecting or handling invalid uses of these ranges.

I like your code (modulo minor stuff) better. PTAL.

File: google/cloud/storage/internal/grpc/object_request_parser.cc [hunk 2]
```diff
@@ -377,30 +377,29 @@ StatusOr<google::storage::v2::ReadObjectRequest> ToProto(
...
     auto const range = request.GetOption<storage::ReadRange>().value();
-    r.set_read_offset(range.begin);
-    r.set_read_limit(range.end - range.begin);
+    auto const merged_offset = (std::max)(range.begin, offset);
+    auto const merged_limit = range.end - merged_offset;
+    if (merged_limit <= 0) {
+      return internal::InvalidArgumentError(
+          absl::StrCat("Invalid gRPC range offset=", merged_offset,
+                       ", limit=", merged_limit, ") from offset=", offset,
+                       " range=[", range.begin, ",", range.end, ")"),
```
Comment (dbolduc) [inline google/cloud/storage/internal/grpc/object_request_parser.cc:392]: Check your brackets. I see two closing parentheses and one opening square bracket.

**Review by dbolduc — APPROVED** [review]

**Pull request merged**